### PR TITLE
Add HasSection to C++ API

### PR DESCRIPTION
**Pull request by evorw**
This function lets you check for the existence of a section, similar to HasValue. We found this useful, hopefully others do too.

### CODE REVIEW

File: cpp/INIReader.cpp [hunk 1]
```diff
@@ -68,6 +68,14 @@ bool INIReader::GetBoolean(const string& section, const string& name, bool defau
         return default_value;
 }
 
+// Return true if the given section exists
```
Comment (benhoyt) [inline cpp/INIReader.cpp]: Let's remove this comment so the doc is just in one place (the header file).

File: cpp/INIReader.cpp [hunk 2]
```diff
@@ -68,6 +68,14 @@ bool INIReader::GetBoolean(const string& section, const string& name, bool defau
...
+bool INIReader::HasSection(const std::string& section) const
```
Comment (benhoyt) [inline cpp/INIReader.cpp]: Let's remove the `std::` prefix here as it's not needed with the `using` at the top (I realize `HasValue` has it, but I should remove that).

```suggestion
bool INIReader::HasSection(const string& section) const
```

File: cpp/INIReader.cpp [hunk 3]
```diff
@@ -68,6 +68,14 @@ bool INIReader::GetBoolean(const string& section, const string& name, bool defau
...
+{
+    const string test = MakeKey(section, "");
+    return std::any_of(_values.begin(), _values.end(),
+        [&](const std::pair<string, string>& k){return k.first.find(test) == 0;});
```
Comment (benhoyt) [inline cpp/INIReader.cpp]: I believe `any_of` loops through all values in the map. This is an ordered map, so we should be able to to a binary search for the position and see if a value with that section exists. How about something like this (not extensively tested):

```c++
bool INIReader::HasSection(const string& section) const
{
    string key = MakeKey(section, "");
    std::map<string, string>::const_iterator pos = _values.lower_bound(key);
    if (pos == _values.end())
        return false;
    // Does the key at the lower_bound pos start with "section"?
    return pos->first.compare(0, key.length(), key) == 0;
}
```

File: cpp/INIReader.h [hunk 4]
```diff
@@ -47,6 +47,9 @@ class INIReader
     // and valid false values are "false", "no", "off", "0" (not case sensitive).
     bool GetBoolean(const std::string& section, const std::string& name, bool default_value) const;
 
+    // Return true if the given section exists
```
Comment (benhoyt) [inline cpp/INIReader.h]:
```suggestion
    // Return true if the given section exists (section must contain at least
    // one name=value pair).
```

File: cpp/INIReader.cpp (hunk 3 above)
Comment (evorw) [reply cpp/INIReader.cpp]: Yep, that's better. I updated the PR.

**Pull request merged**